src/main/java/org/homework/TodoRepository.java

```diff
@@ -1,32 +1,81 @@
 package org.homework;
 
+
+import java.time.LocalDate;
+import java.time.temporal.ChronoUnit;
+import java.util.Comparator;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.homework.Utils.daysBetweenTodayAndDueDate;
+
 
 public class TodoRepository {
     private final Map<Integer, Todo> todoMap = new HashMap<>();
     private int currentId = 0;
 
-    public Todo addTodo(String description) {
+
+    public Todo addTodoWithDueDate(String description, String dueDate) { // Todo 생성 및 저장
         currentId++;
-        Todo newTodo = new Todo(currentId, description);
-        todoMap.put(currentId, newTodo);
+        Todo newTodo = new Todo(currentId, description, dueDate );
+        saveTodo(newTodo);
         return newTodo;
     }
 
+    public void saveTodo(Todo todo) { // Todo 저장
+        todoMap.put(currentId, todo);
+    }
+
     public int checkTodoMapSize(Map<Integer, Todo> todos) {
         return todos.keySet().size();
     }
 
     public Todo findTodoById(int id) {
         return todoMap.get(id);
-    }
+    } // 단건 Todo 불러오기
 
     public void removeTodoById(int id) {
         todoMap.remove(id);
-    }
+    } // Todo 삭제
 
     public Map<Integer, Todo> getAllTodos() {
         return todoMap;
+    } // 전체 Todo 불러오기
+
+
+    public Map<Integer, Todo> getAllTodosWithinDDay(int dDay) {
+        Map<Integer, Todo> resultTodoMap = new HashMap<>();
+        for (Todo todo : todoMap.values()) {
+            if (daysBetweenTodayAndDueDate(todo.getDueDate()) <= dDay) {
+                resultTodoMap.put(todo.getId(), todo);
+            }
+        }
+        return resultTodoMap;
+    }
+
+    public List<Todo> filterAndSortTodos (int dDay) {
+        LocalDate today = LocalDate.now();
+        Map<Integer, Todo> filteredTodos = getAllTodosWithinDDay(dDay);
+        List<Todo> sortedTodos = filteredTodos.values().stream()
+                .filter(todo -> {
+                    LocalDate dueDate = LocalDate.parse(todo.getDueDate());
+                    long daysBetween = ChronoUnit.DAYS.between(today, dueDate);
+                    return daysBetween >= 0 && daysBetween <= dDay;
+                })
+                .sorted(Comparator.comparing(todo -> LocalDate.parse(todo.getDueDate())))
+                .collect(Collectors.toList());
+
+        return sortedTodos;
+    }
+
+    public List<Todo> searchTodosByKeyword(String keyword) {
+        Map<Integer, Todo> todos = getAllTodos();
+        List<Todo> filteredAndSortedTodos = todos.values().stream()
+                .filter(todo -> todo.getDescription().contains(keyword))
+                .sorted(Comparator.comparing(todo -> LocalDate.parse(todo.getDueDate())))
```

Comment:
stream 잘 활용하였습니다.

객체지향에서는 단순 setter,getter를 지양합니다.
todo클래스에게 해당 키워드가 있는지 물어보는 방식으로 코드를 고쳐볼까요?

```suggestion
                .filter(todo -> todo.containsKeyword(keyword))
                .sorted(Comparator.comparing(todo -> LocalDate.parse(todo.getDueDate())))
```